plutus-contract/src/Wallet/Emulator/Wallet.hs

```diff
@@ -264,16 +262,21 @@ handleWallet feeCfg = \case
         utxo <- get >>= ownOutputs
         slotConfig <- WAPI.getClientSlotConfig
         let utx = finalize slotConfig utx'
-        utxWithFees <- validateTxAndAddFees feeCfg slotConfig utxo utx
-        -- balance to add fees
-        tx' <- handleBalanceTx utxo (utx & U.tx . Ledger.fee .~ (utxWithFees ^. U.tx . Ledger.fee))
-        tx'' <- handleAddSignature tx'
-        logInfo $ FinishedBalancing tx''
-        pure $ Right tx''
-
-    walletAddSignatureH :: (Member (State WalletState) effs) => CardanoTx -> Eff effs CardanoTx
-    walletAddSignatureH (Left _) = error "Wallet.Emulator.Wallet.handleWallet: Expecting a mock tx, not an Alonzo tx when adding a signature."
-    walletAddSignatureH (Right tx) = Right <$> handleAddSignature tx
+        -- Balance with dummy fee to get a better estimate of the number of inputs and outputs
+        -- Use a large value otherwise `evaluateTransactionFee` calculates a value 1 or 2 lovelace too few
+        tx <- handleBalanceTx utxo (utx & U.tx . Ledger.fee .~ Ada.lovelaceValueOf 1000000)
+        let requiredSigners = Map.keys (U.unBalancedTxRequiredSignatories utx)
+        theFee <- either (throwError . WAPI.ToCardanoError) pure $ evaluateTransactionFee requiredSigners tx
+        tx' <- handleBalanceTx utxo (utx & U.tx . Ledger.fee .~ theFee)
+        cTx <- either (throwError . WAPI.ToCardanoError) pure $ fromPlutusTx requiredSigners tx'
```

Comment:
This is the only place where this is possible since we have the required signers here.
I have added an explanation in the note that explains why we have the Both constructor.